docs/proposals/20231122-security-groups.md

```diff
@@ -0,0 +1,169 @@
+
+As a cluster operator I want to have more control over my security groups for my Control Plane and Worker Nodes. I want to be able to specify the rules I want to have for my Control Plane and Worker Nodes and not have to rely on the default rules that are currently enforced by CAPO. This will allow me to have more control over my security groups and not have to worry about rules that are not needed by my cluster.
+
+### Implementation Details/Notes/Constraints
+
+The plan is to modify the existing Controllers that already handle opinionated Security Groups when the `managedSecurityGroups` is set to `True`. Until we have the need to create a new Controller, we'll use the existing ones.
+
+An operator will be able to decide how to manage the Security Groups for the Control Plane and Worker Nodes by using a new API. This new API will be used to create the Security Groups and the rules that are needed for the Control Plane and Worker Nodes.
+
+#### Data model changes
+
+`OpenStackCluster` will be changed and the proposed changes are purely additive and nonbreaking. No removals should be required to the data model. For `OpenStackCluster` we will add a new optional field in `spec.SecurityGroupsSpec` and then under it we'll have both `ControlPlaneSecurityGroupRules` and `WorkerSecurityGroupRules` which will represent the Security Groups for the Control Plane and Worker Nodes respectively.
+
+Example `OpenStackCluster` yaml:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: OpenStackCluster
+metadata:
+  name: cluster-sample
+  namespace: default
+  spec:
+    securityGroupsSpec:
+      controlPlaneSecurityGroupRules:
+        - direction: ingress
+          ethertype: IPv4
+          portRangeMin: 22
+          portRangeMax: 22
+          protocol: tcp
+        - direction: ingress
+          ethertype: IPv4
+          portRangeMin: 6443
+          portRangeMax: 6443
+          protocol: tcp
+        - direction: ingress
+          ethertype: IPv4
+          portRangeMin: 2379
+          portRangeMax: 2380
+          protocol: tcp
+        - direction: ingress
+          ethertype: IPv4
+          portRangeMin: 10250
+          portRangeMax: 10250
+          protocol: tcp
+      workerSecurityGroupRules:
+        - direction: ingress
+          ethertype: IPv4
+          portRangeMin: 22
+          portRangeMax: 22
+          protocol: tcp
+        - direction: ingress
+          ethertype: IPv4
+          portRangeMin: 10250
+          portRangeMax: 10250
+          protocol: tcp
+        - direction: ingress
+          ethertype: IPv4
+          portRangeMin: 80
+          portRangeMax: 80
+          protocol: tcp
+        - direction: ingress
+          ethertype: IPv4
+          portRangeMin: 443
+          portRangeMax: 443
+          protocol: tcp
+```
+
+Legacy rules and new rules will be managed exclusively, until we decide to remove the legacy rules.
```

Comment:
If we want to remove the legacy rules, I think we *must* then provide working/minimal defaults for the new managed rules. The user should not need to specify "common sense" rules that would be needed in standard clusters IMO

Reply:
After some discussion with Matt, i've reworked the whole KEP. Please have a look again.